File: server/Sweep/src/impl/data/Template.fs
namespace Sweep.Data

open System
open Sweep.Model.Template
open Sql
open Sweep.Exceptions
open Newtonsoft.Json

module Template = 

  let deserializeTemplate (prop,value:obj) =
   match prop with
    | "Id" ->
      value.ToString() :> obj
    | "SendTo" ->
      JsonConvert.DeserializeObject<string[]> (value.ToString()) :> obj
    | _ -> 
      value

  let add content subject fromName fromAddress sendTo organizationId userId : Template =
    let ctx = GetDataContext()
    let template = ctx.SweepDb.Template.Create()
    let id = Guid.NewGuid().ToString()
    template.Content <- content
    template.Subject <- Some(subject)
    template.FromName <- Some(fromName)
    template.FromAddress <- Some(fromAddress)
    template.SendTo <- JsonConvert.SerializeObject sendTo
    template.OrganizationId <- organizationId
    template.UserId <- userId
    template.Id <- id
    ctx.SubmitUpdates()
    {
      Template.Id = id;
      Content = content;
      Subject = subject;
      FromName = fromName;
      FromAddress = fromAddress;
      SendTo = sendTo;
      OrganizationId = organizationId;
      UserId = userId;
      Deleted = false
    }

  let delete id orgId userId =
    let ctx = GetDataContext()
    let row = query {
      for template in ctx.SweepDb.Template do
      where (template.Id = id && template.OrganizationId = orgId)
      select template
      exactlyOneOrDefault
    } 
    match (isNull row) with 
    | true ->
      raise (NotFoundException("Template not found"))      
    | _ ->
      row.Deleted <- Some((sbyte)1)
      ctx.SubmitUpdates()
               
  let get id organizationId = 
    let ctx = GetDataContext()
    let row = query {      
      for template in ctx.SweepDb.Template do
      where (template.Id = id && template.OrganizationId = organizationId && (template.Deleted.IsNone || (template.Deleted.IsSome && template.Deleted.Value = sbyte(0))))
      select (template)
      exactlyOneOrDefault
    } 
    if isNull row then
      raise (NotFoundException("Not found"))
    else
      row.MapTo<Template>(deserializeTemplate)

  let update id content subject fromName fromAddress sendTo organizationId = 
    let ctx = GetDataContext()
    let row = query {      
      for template in ctx.SweepDb.Template do
      where (template.Id = id && template.OrganizationId = organizationId)
      select (template)
      exactlyOneOrDefault
    } 
    if isNull row then
      raise (NotFoundException("Not found"))
    else
      row.SendTo <- JsonConvert.SerializeObject sendTo
      row.Content <- content
      row.Subject <- Some(subject)
      row.FromAddress <- Some(fromAddress)
      row.FromName <- Some(fromName)
      ctx.SubmitUpdates()
      row.MapTo<Template>(deserializeTemplate)

  let list organizationId =
    let ctx = GetDataContext()
    query {      
      for template in ctx.SweepDb.Template do
      where (template.OrganizationId = organizationId && (template.Deleted.IsNone || (template.Deleted.IsSome && template.Deleted.Value = sbyte(0))))
      select (template)
    } |> Seq.map (fun x -> x.MapTo<Template>(deserializeTemplate))
      |> Seq.toArray

File: clients/lib/dart/lib/model/template.dart
part of openapi.api;

class Template {
  
  String id = null;
  
  String content = null;
  
  String subject = null;
  
  String fromAddress = null;
  
  String fromName = null;
  
  List<String> sendTo = [];
  
  String organizationId = null;
  
  bool deleted = null;
  
  String userId = null;
  Template();

  @override
  String toString() {
    return 'Template[id=$id, content=$content, subject=$subject, fromAddress=$fromAddress, fromName=$fromName, sendTo=$sendTo, organizationId=$organizationId, deleted=$deleted, userId=$userId, ]';
  }

  Template.fromJson(Map<String, dynamic> json) {
    if (json == null) return;
    if (json['id'] == null) {
      id = null;
    } else {
          id = json['id'];
    }
    if (json['content'] == null) {
      content = null;
    } else {
          content = json['content'];
    }
    if (json['subject'] == null) {
      subject = null;
    } else {
          subject = json['subject'];
    }
    if (json['fromAddress'] == null) {
      fromAddress = null;
    } else {
          fromAddress = json['fromAddress'];
    }
    if (json['fromName'] == null) {
      fromName = null;
    } else {
          fromName = json['fromName'];
    }
    if (json['sendTo'] == null) {
      sendTo = null;
    } else {
      sendTo = (json['sendTo'] as List).cast<String>();
    }
    if (json['organizationId'] == null) {
      organizationId = null;
    } else {
          organizationId = json['organizationId'];
    }
    if (json['deleted'] == null) {
      deleted = null;
    } else {
          deleted = json['deleted'];
    }
    if (json['userId'] == null) {
      userId = null;
    } else {
          userId = json['userId'];
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'content': content,
      'subject': subject,
      'fromAddress': fromAddress,
      'fromName': fromName,
      'sendTo': sendTo,
      'organizationId': organizationId,
      'deleted': deleted,
      'userId': userId
    };
  }

  static List<Template> listFromJson(List<dynamic> json) {
    return json == null ? new List<Template>() : json.map((value) => new Template.fromJson(value)).toList();
  }

  static Map<String, Template> mapFromJson(Map<String, dynamic> json) {
    var map = new Map<String, Template>();
    if (json != null && json.length > 0) {
      json.forEach((String key, dynamic value) => map[key] = new Template.fromJson(value));
    }
    return map;
  }
}

